src/main/java/org/opensearch/security/action/apitokens/ApiTokenRepository.java

```diff
@@ -13,19 +13,48 @@
 
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import com.google.common.annotations.VisibleForTesting;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 import org.opensearch.client.Client;
 import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
 import org.opensearch.index.IndexNotFoundException;
 import org.opensearch.security.authtoken.jwt.ExpiringBearerAuthToken;
 import org.opensearch.security.identity.SecurityTokenManager;
 
 public class ApiTokenRepository {
     private final ApiTokenIndexHandler apiTokenIndexHandler;
     private final SecurityTokenManager securityTokenManager;
+    private static final Logger log = LogManager.getLogger(ApiTokenRepository.class);
 
+    private final Map<String, Permissions> jtis = new ConcurrentHashMap<>();
+
+    void reloadApiTokensFromIndex() {
+        jtis.clear();
+        Map<String, ApiToken> tokensFromIndex = apiTokenIndexHandler.getTokenMetadatas();
```

Comment:
I think there's a risk of a race-condition here where `jtis` will be empty for a brief period. 

With the security index, the `configCache` is keyed by `CType` and it does a blanket replacement after reloading: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java#L561

If the same instance of `jtis` is kept, does it make sense to first replace any of the existing keys and remove and keys that don't exist in `tokensFromIndex`, but exist in the `jtis`?

Reply:
I got what you are saying. I switched it such that first I get all tokens from the index, then boot out any ones that don't exist anymore, then add any new entries in: [685dcfc](https://github.com/opensearch-project/security/pull/4992/commits/685dcfc8b7752fc4b118b1b85e3b9e6f827e7c0d). I think this would still result in a short period of time in which revoked api tokens could still be used, but I think it is short enough, and this addresses the original concern. Thoughts @cwperks ?